### DISCUSSION TITLE
HoloLens 2 native C++ and OpenXR Development Models for WebView2

### DISCUSSION THREAD

**Discussion by laultman** [MicrosoftEdge/WebView2Feedback#4769] [Q&A]
The Microsoft documentation of WebView2 has a lot of words but not much in the way of direction outside the "normal" use in Windows. Many developers are working in industry on projects that are not a PC desktop and don't care about the Windows UI. I am one of those that the Microsoft documentation forgets about. MS docs talk about [Windowed vs Visual Hosting of the WebView2](https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/windowed-vs-visual-hosting?tabs=win32cpp) but don't offer any practical information. Should I use this or that.
HoloLens 2 is Windows 11 - but it is NOT in any way a PC!

**My question is this**, considering that I am using HoloLens 2 with OpenXR in a C++ native app and NOT any of the Windows UI frameworks, does this mean that I should use the **CoreWebView2CompositionController** instead of the **CoreWebView2Controller**? (Note: you can see some of my issues here #4768, related to this discussion.)

I am trying to understand, looking for more details to make good decisions. Please offer as much detail as you can. I need to know the reasoning behind the decision process.

**Comment by laultman** [comment]
Still working on this problem. I need some assistance with this confounded error!

TL;DR Background, skip to bottom for the problem.
I decided after a ton of time, to take a completely different approach. I decided to create a completely new project without OpenXR or ARM64 - just c++ on intel on a Windows 11 machine. Do nothing fancy. So, I used the VS template for Core App (C++/WinRT). I did upgrade the Microsoft.Windows.CppWinRT Nuget package. Otherwise, I did not want to bring in anything that certainly will not be used in OpenXR/ARM64 later. I configured the app to use the new C++20 - no sense in backward compatibility - not necessary this is all new code. After some tinkering my app works great on a x64 box with a conventional UWP without the WinUI framework. Note: the CoreWindowsApp offers the IFramework interface that creates a CoreWindow to host your Compositor compositor = m_visuals.Compositor();
I cloned my working code so that I could now add in the WebView2 control. This requires two additional Nuget packages. The Microsoft.Web.WebView2 and the Microsoft.Windows.ImplementationLibrary. You have add a few includes. There is a nasty name collision problem with between the winrt/windows.ui.core.h and corewindow.h. The best way to handle it is by uses fully qualified namespacing.
Problem 1: WebView2 requires the HWND of the UI thread. In this case it is the CoreWindow created by the IFramework (although I would not bet my life on it.) It is not trivial to get the HWND - at least without the special sauce - CoPilot, sorry none of your suggestions worked.
```
    // Get the HWND from the CoreWindow
    HWND GetHWNDFromCoreWindow(CoreWindow const& window)
    {
        // Getting an HWND from the CoreWindow https://kennykerrca.wordpress.com/2012/11/09/windows-8-whered-you-put-my-hwnd/
        // With the window activated, we can now extract the HWND from the CoreWindow
        struct __declspec(uuid("45D64A29-A63E-4CB6-B498-5781D298CB4F")) __declspec(novtable) ICoreWindowInterop : IUnknown {
            virtual HRESULT __stdcall get_WindowHandle(HWND* hwnd) = 0;
        };

        HWND hwnd = nullptr;
        try
        {
            winrt::com_ptr<ICoreWindowInterop> interop{};
            winrt::check_hresult(winrt::get_unknown(window)->QueryInterface(interop.put()));
            winrt::check_hresult(interop->get_WindowHandle(&hwnd)); // Store the HWND in the shared data
        }
        catch (const std::exception&)
        {
            hwnd = nullptr;
        }
        return hwnd;
    }

```
Now with all the prerequisites.  I'm picking up the HWND in the IFramework's `SetWindow(HWND hwnd)` method. There was a note somewhere in all the MS docs about the `window.Activate()` has to happen before trying to hook the [WebView2](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2environment.createcorewebview2controllerasync?view=webview2-dotnet-1.0.2651.64) controller to it. I added an event handler to let me check that the Window is activated. `window.Activated({ this, &App::OnWindowActivated });` I don't think it is necessary but for completeness.

The problem: The WebView2 initialization is straight out of the samples. It fails in the x64 with the error below:
(VS debugger output starting at the call to initialize the WebView2 control)

> Window activated, initialize WV2
***JLA starting InitializeWebView2
'CoreApp2Test.exe' (Win32): Loaded 'C:\Program Files (x86)\Microsoft\EdgeWebView\Application\128.0.2739.42\EBWebView\x64\EmbeddedBrowserWebView.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\Windows.Storage.ApplicationData.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\windows.storage.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\propsys.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\Windows.System.Launcher.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\msvcp110_win.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\shlwapi.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\version.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\ole32.dll'. 
***JLA CreateCoreWebView2EnvironmentWithOptions
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\shell32.dll'. 
***JLA initialized InitializeWebView2.
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\edputil.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\Windows.StateRepositoryClient.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\capauthz.dll'. 
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\ntmarta.dll'. 
**WebView2_SetParent failed: The parameter is incorrect.
SetParent failed while parenting browser window with error: The parameter is incorrect.**
'CoreApp2Test.exe' (Win32): Loaded 'C:\Windows\System32\dbghelp.dll'. 
'CoreApp2Test.exe' (Win32): Unloaded 'C:\Windows\System32\dbghelp.dll'
***JLA failed CreateCoreWebView2Controller
